### ISSUE TITLE
System.AccessViolationException on OpenTK.Wpf.DxGlContext.Dispose()

### ISSUE THREAD

**Issue by simfero** [opentk/GLWpfControl#129]
I upgraded to 4.3.1 from 4.2.3 and now I get the following exception when control is disposed:
System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'
HResult=0x80004003

```
            var settings = new GLWpfControlSettings
            {
                MajorVersion = 3,
                MinorVersion = 1,
                RenderContinuously = false,
                TransparentBackground = true
            };
```

**Comment by NogginBops** [comment]
Could you provide the full stack trace of the exception?

**Comment by NogginBops** [comment]
I think I've found the issue.

We're calling `WGL.DXCloseDevice()` with the wrong handle, it should be `GLDeviceHandle` instead of `DxDevice.Handle` I'm pretty sure.
https://github.com/opentk/GLWpfControl/blob/28d5b99005efcf8bde001db0f48e3f7bb3cb7509/src/GLWpfControl/DXGLContext.cs#L119 

It's weird that my testing hasn't found this problem earlier. Can you confirm what GPU you are using and what driver version is in use?

**Comment by NogginBops** [comment]
I have fixed this issue and verified that it works and the the previous code does indeed crash on my computer as well.
https://github.com/opentk/GLWpfControl/commit/a4706a8ba265222bcd637ef7d5d7dc511b3ec7c1

Ideally I'd like to investigate and fix the integrated graphics card troubles in #128 before releasing `4.2.4`, but not sure how long that will take. But if it takes too long I can release this fix separately.

**Comment by anose001** [comment]
@NogginBops, I ran into the exact same error exception 
> "System.AccessViolationException HResult=0x80004003 Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt." 

in the context of changing between different views in my application. This occured when the view was either implemented as a transient or singleton within the DI setup. After a few clicks and changing views I ended up with the above error exception and the code snippet that you pointed out:

> We're calling `WGL.DXCloseDevice()` with the wrong handle, it should be `GLDeviceHandle` instead of `DxDevice.Handle` I'm pretty sure.
> 
> https://github.com/opentk/GLWpfControl/blob/28d5b99005efcf8bde001db0f48e3f7bb3cb7509/src/GLWpfControl/DXGLContext.cs#L119

**Is there any way for me to test your fix?**

Also, this may or may not be related but when debugging I tried to manually dispose of resources and called GlWpfControl.Dispose(); in the Unload Event and it crashed my application with 

>"The program '[49452] MyProgram.exe' has exited with code 3221225477 (0xc0000005) 'Access violation', 

which at this point I assume is due the wrong handle."

In general, is it good or bad practice of disposing the GlWpfControl when changing views? Or is another approach recommended?

Thank you for all your help!

**Comment by anose001** [comment]
@NogginBops, So I forked repo and added it to my current project in that way. Your code fix seems to resolve the `System.AccessViolationException`. However, within the context of switching views I run into new issues. I will open a separate issue for that.

**Comment by simfero** [comment]
> I have fixed this issue and verified that it works and the the previous code does indeed crash on my computer as well. [a4706a8](https://github.com/opentk/GLWpfControl/commit/a4706a8ba265222bcd637ef7d5d7dc511b3ec7c1)
> 
> Ideally I'd like to investigate and fix the integrated graphics card troubles in #128 before releasing `4.3.2`, but not sure how long that will take. But if it takes too long I can release this fix separately.

@NogginBops I think you should release 4.3.2 with this hotfix since it's a high impact bug.

**Comment by NogginBops** [comment]
@simfero I'll look at this tomorrow. I think I can merge #131 as it is, it fixes some of the issue in #128 so it's probably better to have it than to not have it.
